### PR TITLE
GEOMESA-2984 Support for Scala 2.12

### PR DESCRIPTION
**Pull request by elahrvivaz**
Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>
Co-authored-by: Jim Hughes <jhughes@ccri.com>

### CODE REVIEW

**Comment by elahrvivaz** [comment]
I updated the github actions to run both the 2.11 build and the 2.12 build (also fixed it so that pushes and prs don't cause double runs)

File: build/cqs.tsv [hunk 1]
```diff
@@ -8,10 +8,10 @@ com.clearspring.analytics:stream	2.9.2	compile
...
 com.esotericsoftware:kryo-shaded	3.0.3	compile
 com.esotericsoftware:minlog	1.3.0	compile
-com.fasterxml.jackson.core:jackson-annotations	2.9.10	compile
-com.fasterxml.jackson.core:jackson-core	2.9.10	compile
-com.fasterxml.jackson.core:jackson-databind	2.9.10.4	compile
-com.fasterxml.jackson.dataformat:jackson-dataformat-cbor	2.9.10	compile
+com.fasterxml.jackson.core:jackson-annotations	2.12.1	compile
+com.fasterxml.jackson.core:jackson-core	2.12.1	compile
+com.fasterxml.jackson.core:jackson-databind	2.12.1	compile
+com.fasterxml.jackson.dataformat:jackson-dataformat-cbor	2.12.1	compile
```
Comment (jnh5y) [inline build/cqs.tsv:14]: Have you had a chance to check these at ClearlyDefined.io yet?

Comment (jnh5y) [reply build/cqs.tsv:14]: (That'll let us know if actually need any CQs...)

Comment (elahrvivaz) [reply build/cqs.tsv:14]: no, wanted to double check with you about the version changes first.

File: build/cqs.tsv [hunk 2]
```diff
@@ -249,8 +248,8 @@ org.objenesis:objenesis	1.0	compile
 org.parboiled:parboiled-core	1.2.0	compile
 org.parboiled:parboiled-scala_2.11	1.2.0	compile
 org.postgresql:postgresql	42.2.12	compile
-org.scala-lang.modules:scala-parser-combinators_2.11	1.0.6	compile
-org.scala-lang.modules:scala-xml_2.11	1.0.6	compile
+org.scala-lang.modules:scala-parser-combinators_2.11	1.1.2	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv:251]: https://clearlydefined.io/definitions/maven/mavencentral/org.scala-lang.modules/scala-parser-combinators_2.11/1.1.2

File: build/cqs.tsv [hunk 3]
```diff
@@ -249,8 +248,8 @@ org.objenesis:objenesis	1.0	compile
...
+org.scala-lang.modules:scala-xml_2.11	1.3.0	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv:252]: https://clearlydefined.io/definitions/maven/mavencentral/org.scala-lang.modules/scala-xml_2.11/1.3.0

File: build/cqs.tsv (hunk 1 above)
Comment (elahrvivaz) [reply build/cqs.tsv:14]: i don't see these versions in clearly defined

File: build/cqs.tsv [hunk 4]
```diff
@@ -290,9 +289,9 @@ org.apache.hadoop:hadoop-yarn-common	2.8.5	provided
 org.apache.hbase:hbase-server	2.2.3	provided
 org.apache.kafka:kafka_2.11	2.1.1	provided
 org.apache.kafka:kafka-clients	2.1.1	provided
-org.apache.spark:spark-catalyst_2.11	2.3.1	provided
-org.apache.spark:spark-core_2.11	2.3.1	provided
-org.apache.spark:spark-sql_2.11	2.3.1	provided
+org.apache.spark:spark-catalyst_2.11	2.4.7	provided
```
Comment (elahrvivaz) [inline build/cqs.tsv:292]: https://clearlydefined.io/definitions/maven/mavencentral/org.apache.spark/spark-catalyst_2.11/2.4.7

File: build/cqs.tsv [hunk 5]
```diff
@@ -290,9 +289,9 @@ org.apache.hadoop:hadoop-yarn-common	2.8.5	provided
...
+org.apache.spark:spark-core_2.11	2.4.7	provided
```
Comment (elahrvivaz) [inline build/cqs.tsv:293]: https://clearlydefined.io/definitions/maven/mavencentral/org.apache.spark/spark-core_2.11/2.4.7

File: build/cqs.tsv [hunk 6]
```diff
@@ -290,9 +289,9 @@ org.apache.hadoop:hadoop-yarn-common	2.8.5	provided
...
+org.apache.spark:spark-sql_2.11	2.4.7	provided
```
Comment (elahrvivaz) [inline build/cqs.tsv:294]: https://clearlydefined.io/definitions/maven/mavencentral/org.apache.spark/spark-sql_2.11/2.4.7

File: build/cqs.tsv [hunk 7]
```diff
@@ -8,10 +8,10 @@ com.clearspring.analytics:stream	2.9.2	compile
 com.conversantmedia:disruptor	1.2.13	compile
 com.esotericsoftware:kryo-shaded	3.0.3	compile
 com.esotericsoftware:minlog	1.3.0	compile
-com.fasterxml.jackson.core:jackson-annotations	2.9.10	compile
-com.fasterxml.jackson.core:jackson-core	2.9.10	compile
-com.fasterxml.jackson.core:jackson-databind	2.9.10.4	compile
-com.fasterxml.jackson.dataformat:jackson-dataformat-cbor	2.9.10	compile
+com.fasterxml.jackson.core:jackson-annotations	2.12.1	compile
+com.fasterxml.jackson.core:jackson-core	2.12.1	compile
+com.fasterxml.jackson.core:jackson-databind	2.12.1	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv:13]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22967

File: build/cqs.tsv [hunk 8]
```diff
@@ -8,10 +8,10 @@ com.clearspring.analytics:stream	2.9.2	compile
 com.conversantmedia:disruptor	1.2.13	compile
 com.esotericsoftware:kryo-shaded	3.0.3	compile
 com.esotericsoftware:minlog	1.3.0	compile
-com.fasterxml.jackson.core:jackson-annotations	2.9.10	compile
-com.fasterxml.jackson.core:jackson-core	2.9.10	compile
-com.fasterxml.jackson.core:jackson-databind	2.9.10.4	compile
-com.fasterxml.jackson.dataformat:jackson-dataformat-cbor	2.9.10	compile
+com.fasterxml.jackson.core:jackson-annotations	2.12.1	compile
+com.fasterxml.jackson.core:jackson-core	2.12.1	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv:12]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22966

File: build/cqs.tsv (hunk 1 above)
Comment (elahrvivaz) [inline build/cqs.tsv:14]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22968

**Comment by elahrvivaz** [comment]
@jnh5y ready for merge

File: pom.xml [hunk 9]
```diff
@@ -71,8 +71,8 @@
         <avro.version>1.8.2</avro.version> <!-- see also confluent.avro.version -->
         <parquet.version>1.9.0</parquet.version>
         <orc.version>1.5.4</orc.version>
-        <fasterxml.jackson.version>2.9.10</fasterxml.jackson.version> <!-- used by geoserver (2.9.7), spark (2.6.7) and arrow (2.7.1) -->
-        <fasterxml.jackson.databind.version>2.9.10.4</fasterxml.jackson.databind.version> <!-- used by spark (2.6.7.1) and arrow (2.7.1) -->
+        <fasterxml.jackson.version>2.12.1</fasterxml.jackson.version> <!-- used by geoserver (2.9.7), spark (2.6.7) and arrow (2.7.1) -->
+        <fasterxml.jackson.databind.version>2.12.1</fasterxml.jackson.databind.version> <!-- used by spark (2.6.7.1) and arrow (2.7.1) -->
```
Comment (elahrvivaz) [inline pom.xml:75]:
```suggestion
        <fasterxml.jackson.databind.version>2.12.1</fasterxml.jackson.databind.version> <!-- used by geoserver (2.9.10.4), spark (2.6.7.1) and arrow (2.7.1) -->
```

File: pom.xml [hunk 10]
```diff
@@ -71,8 +71,8 @@
         <avro.version>1.8.2</avro.version> <!-- see also confluent.avro.version -->
         <parquet.version>1.9.0</parquet.version>
         <orc.version>1.5.4</orc.version>
-        <fasterxml.jackson.version>2.9.10</fasterxml.jackson.version> <!-- used by geoserver (2.9.7), spark (2.6.7) and arrow (2.7.1) -->
-        <fasterxml.jackson.databind.version>2.9.10.4</fasterxml.jackson.databind.version> <!-- used by spark (2.6.7.1) and arrow (2.7.1) -->
+        <fasterxml.jackson.version>2.12.1</fasterxml.jackson.version> <!-- used by geoserver (2.9.7), spark (2.6.7) and arrow (2.7.1) -->
```
Comment (elahrvivaz) [inline pom.xml:74]:
```suggestion
        <fasterxml.jackson.version>2.12.1</fasterxml.jackson.version> <!-- used by geoserver (2.9.10), spark (2.6.7) and arrow (2.7.1) -->
```

**Comment by pomadchin** [comment]
🎉

File: benchmarks/pom.xml [hunk 11]
```diff
@@ -5,7 +5,7 @@
   <parent>
     <groupId>org.locationtech.geomesa</groupId>
     <artifactId>geomesa_2.11</artifactId>
-    <version>3.1.0-SNAPSHOT</version>
+    <version>3.2.0-SNAPSHOT</version>
```
Comment (jnh5y) [inline benchmarks/pom.xml:8]: Hmm.... is behind a profile?

File: build/cqs.tsv (hunk 1 above)
Comment (jnh5y) [inline build/cqs.tsv:14]: I feel like GeoServer and/or Spark uses these Jackson libraries.  Have you tried this PR with either Scala version in GeoServer yet?

File: build/cqs.tsv (hunk 6 above)
Comment (jnh5y) [inline build/cqs.tsv:294]: Do we need to change the minimum supported Spark version?

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/utils/GeoMessageSerializer.scala [hunk 12]
```diff
@@ -111,7 +111,7 @@ object GeoMessageSerializer {
       try {
         // use the feature id if available, otherwise (for clear) use random shard
         if (keyBytes.length > 0) {
-          Math.abs(MurmurHash3.bytesHash(keyBytes)) % count
+          Math.abs(MurmurHash3.bytesHash(keyBytes, MurmurHash3.arraySeed)) % count
```
Comment (jnh5y) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/utils/GeoMessageSerializer.scala:114]: Hmm... why did this change?  (Mainly curious)

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geohash/GeohashUtils.scala [hunk 13]
```diff
@@ -59,12 +59,12 @@ object GeohashUtils
    */
   case class ResolutionRange(minBitsResolution:Int=5,
                              maxBitsResolution:Int=63,
-                             numBitsIncrement:Int=2)
-    extends Inclusive(minBitsResolution, maxBitsResolution, numBitsIncrement) {
+                             numBitsIncrement:Int=2) {
```
Comment (jnh5y) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geohash/GeohashUtils.scala:62]: Hmm... any chance this represents an API change we care about?

File: pom.xml [hunk 14]
```diff
@@ -2755,12 +2755,13 @@
                                 <bannedDependencies>
                                     <excludes combine.children="append">
                                         <exclude>*:*_2.10</exclude>
+                                        <exclude>*:*_2.12</exclude>
```
Comment (jnh5y) [inline pom.xml:2758]: Does the change version script switch this over to _2.11?  (And hence do the right thing?)

File: benchmarks/pom.xml (hunk 11 above)
Comment (elahrvivaz) [reply benchmarks/pom.xml:8]: yeah, we should update the release docs to include it

Comment (elahrvivaz) [reply benchmarks/pom.xml:8]: i'll do that

File: build/cqs.tsv (hunk 1 above)
Comment (elahrvivaz) [reply build/cqs.tsv:14]: i believe so, but it's been a while haha. i think these are not shipped into geoserver. the issue was with the spark unit tests dying on the older jackson versions for some reason, even though the spark version did not change.

File: build/cqs.tsv (hunk 6 above)
Comment (elahrvivaz) [reply build/cqs.tsv:294]: no, it should still work with 2.3.x (although i did not try it). there are no 2.3.x_2.12 spark artifacts iirc so i had to bump it.

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/utils/GeoMessageSerializer.scala (hunk 12 above)
Comment (elahrvivaz) [reply geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/utils/GeoMessageSerializer.scala:114]: scala changed the implementation of this bytes hash in 2.11.10-ish... which would mean that hashes don't match any more, potentially causing duplicated data. calling it this way mimics the old behavior so keeps the hashes consistent across scala versions.

Comment (elahrvivaz) [reply geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/utils/GeoMessageSerializer.scala:114]: this only turned up due to a random failing unit test, and was loads of fun to track down :P

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geohash/GeohashUtils.scala (hunk 13 above)
Comment (elahrvivaz) [reply geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geohash/GeohashUtils.scala:62]: i would say no... `Inclusive` is final in scala 2.12 so we can't extend it any more, hence the change.

File: pom.xml (hunk 14 above)
Comment (elahrvivaz) [reply pom.xml:2758]: yes!

Comment (elahrvivaz) [reply pom.xml:2758]: https://github.com/locationtech/geomesa/pull/2682/files#diff-426d95ee0ba39aed84d2060724c557dff39626437bf23ef0dca67d7fdb7fa6f8R77

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/utils/GeoMessageSerializer.scala (hunk 12 above)
Comment (jnh5y) [reply geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/utils/GeoMessageSerializer.scala:114]: Good job!  Better us tracking it down in a unit test now....;)

**Pull request merged**